dev/com.ibm.ws.webcontainer/src/com/ibm/ws/webcontainer/srt/SRTServletRequest.java

```diff
@@ -2755,7 +2753,7 @@ protected void cleanupFromFinish() {
         this._request.clearHeaders();
         this._request = null; // as SRTServletResponse.finish() does for _response
         this._requestContext.finish();
-        SRTServletRequestThreadData.getInstance().init(null);
+        threadData.init(null);
```

Comment:
should threadData be nulled out?

Reply:
I don't know ... it makes sense, and was how I initially coded this, but the async test failed and I thought maybe this was the cause. But the async test still fails without nulling threadData, so that was not the problem.  I guess we should null out the local reference, after nulling the threadLocal fields (i.e. after the threadData.init(null); above)

Reply:
... and that doesn't work  :-) 
for one request at a time it is OK, but under load it gets NPEs like
`Exception = java.lang.NullPointerException
Source = com.ibm.ws.webcontainer.srt31.SRTConnectionContext.finishConnection
probeid = 74
Stack Dump = java.lang.NullPointerException
        at com.ibm.ws.webcontainer.srt.SRTServletRequest.getDispatchContext(SRTServletRequest.java:4065)
        at com.ibm.ws.webcontainer31.srt.SRTServletRequest31.cleanupFromFinish(SRTServletRequest31.java:151)
        at com.ibm.ws.webcontainer.srt.SRTServletRequest.finish(SRTServletRequest.java:2746)
        at com.ibm.ws.webcontainer31.osgi.srt.SRTConnectionContext31.finishConnection(SRTConnectionContext31.java:216)
        at com.ibm.ws.webcontainer.WebContainer.handleRequest(WebContainer.java:1166)
`
so now I (think I) know the answer - threadData should _not_ be nulled out

Reply:
The reason is the super call.  You can move:
```
        WebApp31 webapp = (WebApp31) ((WebAppDispatcherContext) this.getDispatchContext()).getWebApp();
```
before the call to super in order to get the dispatch context before clearing out the threadData object.

Reply:
Good catch, thanks - that works. Done.

Reply:
as discussed in slack this morning:  

I have realized that we should not null out threadData, because it is replacing calls to SRTServletRequestThreadData.getInstance()  which never returns null  ... it might return a blank slate with no useful data, but there it always gives you back a reference to an SRTServletRequestThreadData object

so I've removed the `threadData = null` here, and reverted the change made in SRTServletRequest31 in cleanupFromFinish() where setting threadData null caused an NPE before, as discussed above